types/src/network_address/mod.rs

```diff
@@ -192,26 +207,119 @@ pub struct EmptyError;
 ////////////////////
 // NetworkAddress //
 ////////////////////
+// Network
+// Transpot
+// Session - Noise
+// Presentation - Handshake
+
+fn is_network_layer(p: Option<&Protocol>, allow_empty: bool) -> bool {
+    use Protocol::*;
+    match p {
+        None => allow_empty,
+        Some(Ip4(_)) => true,
+        Some(Ip6(_)) => true,
+        Some(Dns(_)) => true,
+        Some(Dns4(_)) => true,
+        Some(Dns6(_)) => true,
+        Some(Memory(_)) => cfg!(test),
+        _ => false,
+    }
+}
+
+fn is_transport_layer(p: Option<&Protocol>, allow_empty: bool) -> bool {
+    use Protocol::*;
+    match p {
+        None => allow_empty,
+        Some(Tcp(_)) => true,
+        _ => false,
+    }
+}
+
+fn is_session_layer(p: Option<&Protocol>, allow_empty: bool) -> bool {
+    use Protocol::*;
+    match p {
+        None => allow_empty,
+        Some(NoiseIK(_)) => true,
+        _ => false,
+    }
+}
+
+fn is_handshake_layer(p: Option<&Protocol>, allow_empty: bool) -> bool {
+    use Protocol::*;
+    match p {
+        None => allow_empty,
+        Some(Handshake(_)) => true,
+        _ => false,
+    }
+}
 
 impl NetworkAddress {
-    fn new(protocols: Vec<Protocol>) -> Self {
-        Self(protocols)
+    fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn from_protocols(protocols: Vec<Protocol>) -> Result<Self, ParseError> {
+        use Protocol::*;
+
+        let mut iter = protocols.iter();
+
+        let mut p = iter.next();
+
+        if p == None {
+            return Ok(Self(protocols));
+        }
```

Comment:
This is interesting that you stop along the way, so it is valid to have a shortened address e.g. "", "/dns/example.com", "/dns/example.com/tcp/8080", etc.

What's the reasoning for that?